### PR TITLE
Handshake and internal processing for socket groups

### PR DESCRIPTION
**Pull request by ethouris**

### CODE REVIEW

File: srtcore/api.cpp [hunk 1]
```diff
@@ -746,6 +746,8 @@ int CUDTUnited::newConnection(const SRTSOCKET listen, const sockaddr_any& peer,
 
       CTimer::triggerEvent();
 
+      CTimer::triggerEvent();
+
```
Comment (maxsharabayko) [inline srtcore/api.cpp]: ?

File: srtcore/buffer.h [hunk 2]
```diff
@@ -440,7 +440,7 @@ class CRcvBuffer
        return m_iStartPos == m_iLastAckPos;
    }
    bool full() const { return m_iStartPos == (m_iLastAckPos+1)%m_iSize; }
-   int capacity() const { return m_iSize; }
+   int capacity() const { return m_iSize; } // XXX WHY it was changed to m_iSize-1 ?
```
Comment (maxsharabayko) [inline srtcore/buffer.h]: > // XXX WHY it was changed to m_iSize-1 ?
This comment was already removed from one of the previous PRs.

File: srtcore/packet.cpp [hunk 3]
```diff
@@ -299,6 +300,11 @@ void CPacket::pack(UDTMessageType pkttype, const int32_t* lparam, void* rparam,
       break;
 
    case UMSG_KEEPALIVE: //0001 - Keep-alive
+      if (lparam)
+      {
+          // XXX EXPERIMENTAL. Pass the 32-bit integer here.
+         m_nHeader[SRT_PH_MSGNO] = *lparam;
```
Comment (maxsharabayko) [inline srtcore/packet.cpp]: alignment

File: srtcore/core.cpp [hunk 4]
```diff
@@ -2969,6 +3064,32 @@ bool CUDT::interpretSrtHandshake(const CHandShake& hs,
...
+            else if ( cmd == SRT_CMD_GROUP )
+            {
+                // Note that this will fire in both cases:
+                // - When receiving HS request from the Initiator, which belongs to a group, and agent must
+                //   create the mirror group on his side (or join the existing one, if there's already
+                //   a mirror group for that group ID).
+                // - When receiving HS response from the Responder, with its mirror group ID, so the agent
+                //   must put the group into his peer group data
+                int32_t groupdata[GRPD__SIZE];
+                if ( bytelen < GRPD__SIZE * GRPD_FIELD_SIZE)
```
Comment (maxsharabayko) [inline srtcore/core.cpp]: extra space after the left brace

File: srtcore/core.cpp [hunk 5]
```diff
@@ -3105,6 +3240,385 @@ bool CUDT::checkApplyFilterConfig(const std::string &confstr)
     return true;
 }
 
+bool CUDT::interpretGroup(const int32_t groupdata[], int hsreq_type_cmd SRT_ATR_UNUSED)
+{
+    SRTSOCKET grpid = groupdata[GRPD_GROUPID];
+    SRT_GROUP_TYPE gtp = SRT_GROUP_TYPE(groupdata[GRPD_GROUPTYPE]);
+    //SRTSOCKET master_peerid = groupdata[GRPD_MASTERID];
+    //int32_t tdiff = groupdata[GRPD_MASTERTDIFF];
```
Comment (maxsharabayko) [inline srtcore/core.cpp]: delete commented code

File: srtcore/core.cpp [hunk 6]
```diff
@@ -3105,6 +3240,385 @@ bool CUDT::checkApplyFilterConfig(const std::string &confstr)
...
+
+    /*
+
+    // Synchronize the TSBPD PEER start time with the existing connection,
+    // if there exists the connection with given peer.
+    if (master_peerid != -1)
+    {
+        // Here "I am a peer", so this is the socket ID of local socket of a parallel connection.
+        // Check if it exists, if not, reject the connection.
+        CUDTSocket* master = s_UDTUnited.locateSocket(master_peerid, s_UDTUnited.ERH_RETURN);
```
Comment (maxsharabayko) [inline srtcore/core.cpp]: Delete commented code

File: srtcore/core.cpp [hunk 7]
```diff
@@ -3105,6 +3240,385 @@ bool CUDT::checkApplyFilterConfig(const std::string &confstr)
...
+    else
+    {
+        HLOGC(mglog.Debug, log << "synchronizeWithGroup: @" << m_SocketID
+                << " DEFINED ISN: RCV=%" << m_iRcvLastAck
+                << " SND=%" << m_iSndLastAck);
+    }
+}
+
+// [[using locked(this->m_GroupLock)]];
+bool CUDTGroup::getBufferTimeBase(CUDT* forthesakeof, steady_clock::time_point& w_tb, bool& w_wp, steady_clock::duration& w_dr)
```
Comment (maxsharabayko) [inline srtcore/core.cpp]: `forthesakeof` is really not the best name for this.

File: srtcore/core.cpp [hunk 8]
```diff
@@ -6983,6 +7573,13 @@ int32_t CUDT::ackDataUpTo(int32_t ack)
     if (acksize > 0)
     {
         const int distance = m_pRcvBuffer->ackData(acksize);
+/*
+        XXX example code - inform the group up to which packet
+        data are received so that it can be also acknowledged
+        in all others and false lossreport prevened
+        if (m_parent->m_IncludedGroup)
+            m_parent->m_IncludedGroup->readyPackets(this, ack);
+*/
```
Comment (maxsharabayko) [inline srtcore/core.cpp]: Align `/*` and `*/` with the main code block.

File: srtcore/core.cpp [hunk 9]
```diff
@@ -7100,7 +7697,7 @@ void CUDT::sendCtrl(UDTMessageType pkttype, const int32_t* lparam, void* rparam,
         // IF ack %> m_iRcvLastAck
         if (CSeqNo::seqcmp(ack, m_iRcvLastAck) > 0)
         {
-            ackDataUpTo(ack);
+            int32_t first_seq = ackDataUpTo(ack);
             leaveCS(m_RcvBufferLock);
```
Comment (maxsharabayko) [inline srtcore/core.cpp]: const int32_t first_seq

File: srtcore/core.cpp [hunk 10]
```diff
@@ -7766,7 +8374,14 @@ void CUDT::processCtrl(const CPacket &ctrlpkt)
         // inaccurate. Additionally it won't lock if TSBPD mode is off, and
         // won't update anything. Note that if you set TSBPD mode and use
         // srt_recvfile (which doesn't make any sense), you'll have a deadlock.
-        m_pRcvBuffer->addRcvTsbPdDriftSample(ctrlpkt.getMsgTimeStamp(), m_RecvLock);
+        steady_clock::duration udrift;
+        steady_clock::time_point newtimebase;
+        bool drift_updated = m_pRcvBuffer->addRcvTsbPdDriftSample(ctrlpkt.getMsgTimeStamp(), m_RecvLock,
+                (udrift), (newtimebase));
```
Comment (maxsharabayko) [inline srtcore/core.cpp]: const bool drift_updated

File: srtcore/core.cpp [hunk 11]
```diff
@@ -8243,10 +8870,63 @@ std::pair<int, steady_clock::time_point> CUDT::packData(CPacket& w_packet)
...
+                    {
+                        int32_t seqpair[2];
+                        seqpair[0] = m_iSndCurrSeqNo;
+                        seqpair[1] = w_packet.m_iSeqNo;
+                        HLOGC(mglog.Debug, log << "... sending INITIAL DROP (ISN FIX): "
+                                << "msg=" << MSGNO_SEQ::unwrap(w_packet.m_iMsgNo) << " SEQ:"
+                                << seqpair[0] << " - " << seqpair[1] << "(" << seqdiff << " packets)");
+#ifndef SRT_TEST_DISABLE_KEY_CONTROL_PACKETS
+                        sendCtrl(UMSG_DROPREQ, &w_packet.m_iMsgNo, seqpair, sizeof(seqpair));
+#endif
```
Comment (maxsharabayko) [inline srtcore/core.cpp]: This was already removed from one of the previous PRs.

File: srtcore/core.cpp [hunk 12]
```diff
@@ -8449,23 +9129,73 @@ void CUDT::sendLossReport(const std::vector<std::pair<int32_t, int32_t> > &loss_
...
-   // Just heard from the peer, reset the expiration count.
-   m_iEXPCount = 1;
-   m_tsLastRspTime = steady_clock::now();
+    // XXX This should be called (exclusively) here:
+    // m_pRcvBuffer->addLocalTsbPdDriftSample(packet.getMsgTimeStamp());
+    // Just heard from the peer, reset the expiration count.
+    m_iEXPCount = 1;
+    m_tsLastRspTime = steady_clock::now();
+
+    bool need_tsbpd = m_bTsbPd || m_bGroupTsbPd;
```
Comment (maxsharabayko) [inline srtcore/core.cpp]: const bool need_tsbpd

File: srtcore/core.cpp [hunk 13]
```diff
@@ -8795,6 +9527,10 @@ int CUDT::processData(CUnit* in_unit)
             // Otherwise it's an error.
             if (adding_successful)
             {
+                // XXX move this code do CUDT::defaultPacketArrival and call it from here:
+
+                // srt_loss_seqs = CALLBACK_CALL(m_cbPacketArrival, rpkt);
+
```
Comment (maxsharabayko) [inline srtcore/core.cpp]: Not sure whether this comment should stay, but at least no need in extra blank lines

File: srtcore/core.cpp [hunk 14]
```diff
@@ -9030,6 +9766,70 @@ int CUDT::processData(CUnit* in_unit)
...
+        {
+            HLOGC(mglog.Debug, log << "defaultPacketArrival: IN-GROUP rcv state transition to RUNNING. NOT checking for loss");
+            gi->rcvstate = CUDTGroup::GST_RUNNING;
+            return output;
+        }
+    }
+
+    int initial_loss_ttl = 0;
+    if ( self->m_bPeerRexmitFlag )
+        initial_loss_ttl = self->m_iReorderTolerance;
```
Comment (maxsharabayko) [inline srtcore/core.cpp]:
```
const int initial_loss_ttl = (self->m_bPeerRexmitFlag)
    ? 0
    : self->m_iReorderTolerance;
```

File: srtcore/core.cpp [hunk 15]
```diff
@@ -9030,6 +9766,70 @@ int CUDT::processData(CUnit* in_unit)
...
+    int seqdiff = CSeqNo::seqcmp(pkt.m_iSeqNo, self->m_iRcvCurrSeqNo);
+
+    HLOGC(mglog.Debug, log << "defaultPacketArrival: checking sequence " << pkt.m_iSeqNo
+            << " against latest " << self->m_iRcvCurrSeqNo << " (distance: " << seqdiff << ")");
+
+    // Loss detection.
+    if (seqdiff > 1) // packet is later than the very subsequent packet
+    {
+        int32_t seqlo = CSeqNo::incseq(self->m_iRcvCurrSeqNo);
+        int32_t seqhi = CSeqNo::decseq(pkt.m_iSeqNo);
```
Comment (maxsharabayko) [inline srtcore/core.cpp]: const

File: srtcore/buffer.h (hunk 2 above)
Comment (ethouris) [reply srtcore/buffer.h]: Ok, I'll remove this. We'll research during the reimplementation

File: srtcore/core.cpp (hunk 7 above)
Comment (ethouris) [reply srtcore/core.cpp]: This is a socket core, for which the synchronization is being made. Data prepared and extracted by this function will be then used to set on this one entity.

File: srtcore/core.cpp (hunk 11 above)
Comment (ethouris) [reply srtcore/core.cpp]: Merging again...

File: srtcore/core.cpp (hunk 13 above)
Comment (ethouris) [reply srtcore/core.cpp]: Should stay. This is where the parallel packet lossreport prevention should be added. It's an important improvement for broadcast groups.

**Comment by gou4shi1** [comment]
Hi @ethouris , may I ask you a question about these codes?
```
                // Do this checking only for groups and only at the very first moment,
                // when there's still nothing in the buffer. Otherwise there will be
                // a serious data discrepancy between the agent and the peer.
                // After increasing by 1, but being previously set as ISN-1, this should be == ISN,
                // if this is the very first packet to send.
#if ENABLE_EXPERIMENTAL_BONDING
                // Fortunately here is only the procedure that verifies if the extraction
                // sequence is moved due to the difference between ISN caught during the existing
                // transmission and the first sequence possible to be used at the first sending
                // instruction. The group itself isn't being accessed.
                if (m_parent->m_GroupOf && m_iSndCurrSeqNo != w_packet.m_iSeqNo && m_iSndCurrSeqNo == m_iISN)
                {
                    const int packetspan = CSeqNo::seqcmp(w_packet.m_iSeqNo, m_iSndCurrSeqNo);

                    HLOGC(qslog.Debug, log << CONID() << "packData: Fixing EXTRACTION sequence " << m_iSndCurrSeqNo
                            << " from SCHEDULING sequence " << w_packet.m_iSeqNo
                            << " DIFF: " << packetspan << " STAMP:" << BufferStamp(w_packet.m_pcData, w_packet.getLength()));

                    // This is the very first packet to be sent; so there's nothing in
                    // the sending buffer yet, and therefore we are in a situation as just
                    // after connection. No packets in the buffer, no packets are sent,
                    // no ACK to be awaited. We can screw up all the variables that are
                    // initialized from ISN just after connection.
                    //
                    // Additionally send the drop request to the peer so that it
                    // won't stupidly request the packets to be retransmitted.
                    // Don't do it if the difference isn't positive or exceeds the threshold.
                    if (packetspan > 0)
                    {
                        int32_t seqpair[2];
                        seqpair[0] = m_iSndCurrSeqNo;
                        seqpair[1] = w_packet.m_iSeqNo;
                        HLOGC(qslog.Debug, log << "... sending INITIAL DROP (ISN FIX): "
                                << "msg=" << MSGNO_SEQ::unwrap(w_packet.m_iMsgNo) << " SEQ:"
                                << seqpair[0] << " - " << seqpair[1] << "(" << packetspan << " packets)");
                        sendCtrl(UMSG_DROPREQ, &w_packet.m_iMsgNo, seqpair, sizeof(seqpair));

                        // In case when this message is lost, the peer will still get the
                        // UMSG_DROPREQ message when the agent realizes that the requested
                        // packet are not present in the buffer (preadte the send buffer).
                    }
                }
                else
#endif
                {
                    HLOGC(qslog.Debug, log << CONID() << "packData: Applying EXTRACTION sequence " << m_iSndCurrSeqNo
                            << " over SCHEDULING sequence " << w_packet.m_iSeqNo
                            << " DIFF: " << CSeqNo::seqcmp(m_iSndCurrSeqNo, w_packet.m_iSeqNo)
                            << " STAMP:" << BufferStamp(w_packet.m_pcData, w_packet.getLength()));

#if ENABLE_EXPERIMENTAL_BONDING
                    HLOGC(qslog.Debug, log << "... CONDITION: IN GROUP: " << (m_parent->m_GroupOf ? "yes":"no")
                            << " extraction-seq=" << m_iSndCurrSeqNo << " scheduling-seq=" << w_packet.m_iSeqNo << " ISN=" << m_iISN);
#endif

                    // Do this always when not in a group, 
                    w_packet.m_iSeqNo = m_iSndCurrSeqNo;
                }

```
The logging says `Fixing EXTRACTION sequence m_iSndCurrSeqNo from SCHEDULING sequence w_packet.m_iSeqNo`, but there is no `m_iSndCurrSeqNo = w_packet.m_iSeqNo;`, and all following packets to send will be override by `m_iSndCurrSeqNo` which is starting from wrong ISN.
Moreover, it's sending `w_packet.m_iSeqNo`, why the DROPREQ includes `w_packet.m_iSeqNo`?

I guess the correct code should be like this?
```
sendDropReq(m_iSndCurrSeqNo, w_packet.m_iSeqNo - 1);
m_iSndCurrSeqNo = w_packet.m_iSeqNo
```
Correct me if I'm wrong, thanks!

**Comment by ethouris** [comment]
Ok, I'm not sure what this exactly is without more time spent on the analysis, but what I can see here is that the value `w_packet.m_iSeqNo` is later overridden with the scheduling sequence, while the value used for dropping is the previous one, which is likely the increased ISN. Note that this overriding and dropping happens only if the scheduling sequence and initial sequence differ.

The problem here was that in UDT the idea was that the packet doesn't get the sequence number at the time when it's being scheduled for sending; this number is given to the packet only at the sending queue. It's not even set when inserting into the queue, it's given to it at the time when the next packet is being picked up for sending. This was a problem for the group implementation that has been created on top of socket sending (let's say, without a strong refax of the code, in this case the sending facilities and the sender buffer, it wasn't possible to do any other way). And as a single packet had to be sent over possibly multiple connections at once, it should have the same sequence number on all member connections. Therefore there had to be one central place where this value has to be controlled, and as it was the group, it must have been decided before you call the single-socket sending procedure. It was expected though that the sequence numbers will roll in both systems the same way, which means that the only way how you can have a different "scheduling sequence" and "internal sequence" was the very first packet sending after activation of an idle link.

**Comment by gou4shi1** [comment]
@ethouris Thanks for you explanation.
> what I can see here is that the value w_packet.m_iSeqNo is later overridden with the scheduling sequence

"extraction sequence" should be `m_iSndCurrSeqNo`, while "scheduling sequence" should be `w_packet.m_iSeqNo` which is set in `sendmsg2()`?
Where is the code of "the value w_packet.m_iSeqNo is later overridden with the scheduling sequence" did you mean?

**Comment by ethouris** [comment]
Ups, maybe actuallly I should have spent more time on the analysis ;). 

Ok, it's done only as a replacement for previous "increasing the sequence by 1", so not in a situation when these sequences differ. Here the already set sequence is taken as a good deal. Whether the sequence of the packet being sent currently will be effectively dropped, is another matter. At least what I can see is likely that it will be, as dropping uses simply a pair of two sequences and the message is treated as if a packet was received previously with the sequence being the later one.

If you have that case, you should see in the logs the packet sent after this `packData` call on the receiver side reported as "belated".

**Comment by gou4shi1** [comment]
If the `extraction sequence` is differ from `scheduling sequence` for the first packet, we should override the `extraction sequence` with `scheduling sequence`?

**Comment by ethouris** [comment]
Search for the instruction that assigns to this field. It's about 10 lines down.

**Comment by gou4shi1** [comment]
> Search for the instruction that assigns to this field. It's about 10 lines down.

Did you mean `w_packet.m_iSeqNo = m_iSndCurrSeqNo;`?
But the log is saying `Applying EXTRACTION sequence m_iSndCurrSeqNo over SCHEDULING sequence w_packet.m_iSeqNo`, 
instead of 
> If the `extraction sequence` is differ from `scheduling sequence` for the first packet, we should override the `extraction sequence` with `scheduling sequence`? But I can't see any code that is doing this.

Sorry maybe I understand sth wrong.

**Comment by gou4shi1** [comment]
> I guess the correct code should be like this?

```
diff --git a/srtcore/core.cpp b/srtcore/core.cpp
index 07c4427e..4a906ef1 100644
--- a/srtcore/core.cpp
+++ b/srtcore/core.cpp
@@ -9608,16 +9608,18 @@ bool srt::CUDT::packUniqueData(CPacket& w_packet, time_point& w_origintime)
         {
             int32_t seqpair[2];
             seqpair[0] = m_iSndCurrSeqNo;
-            seqpair[1] = w_packet.m_iSeqNo;
+            seqpair[1] = CSeqNo::decseq(w_packet.m_iSeqNo);
+            const int32_t no_msgno = 0;
             HLOGC(qslog.Debug, log << "... sending INITIAL DROP (ISN FIX): "
                     << "msg=" << MSGNO_SEQ::unwrap(w_packet.m_iMsgNo) << " SEQ:"
                     << seqpair[0] << " - " << seqpair[1] << "(" << packetspan << " packets)");
-            sendCtrl(UMSG_DROPREQ, &w_packet.m_iMsgNo, seqpair, sizeof(seqpair));
+            sendCtrl(UMSG_DROPREQ, &no_msgno, seqpair, sizeof(seqpair));
 
             // In case when this message is lost, the peer will still get the
             // UMSG_DROPREQ message when the agent realizes that the requested
             // packet are not present in the buffer (preadte the send buffer).
         }
+        m_iSndCurrSeqNo = w_packet.m_iSeqNo;
     }
     else
 #endif
```

> Search for the instruction that assigns to this field. It's about 10 lines down.

Or did you mean `overrideSndSeqNo()`? This should only be called in `sendBroadcast()` or `sendBackup()`?

**Comment by ethouris** [comment]
Let me try to start over.

The "extraction sequence" (the value written in `m_iSndCurrSeqNo`) shall be overridden by the scheduling sequence (value written in the packet). In case of a group connection the former can be slightly outdated, although it is believed that it's just few packets behind the scheduling sequence. The scheduling sequence should be the one taking over because this is the value with which this same packet has been sent over another member connection. And all packets before this one should be skipped by the receiver.

The extraction sequence should be used always in case when this is a socket connection, and it gets increased a few lines above.

Whether this upper value of the sequence should be decreased, I'm not sure. It will be then used in `CRcvBufferNew::dropMessage` and `CUDT::dropFromLossLists`  (on the other side). Basing on what's in `CRcvLossList::remove` you are probably right, but speak to Max to make sure.

**Pull request merged**